src/Passes/Source/Generators/LlvmPassesConfig.hpp

```diff
@@ -0,0 +1,29 @@
+#pragma once
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "Commandline/ConfigurationManager.hpp"
+
+namespace microsoft
+{
+namespace quantum
+{
+
+    class LlvmPassesConfiguration
+    {
+      public:
+        void setup(ConfigurationManager& config);
+
+        static LlvmPassesConfiguration disable();
+        bool                           alwaysInline() const;
+
+        bool isDisabled() const;
```

Comment:
My understanding is that this is a shared configuration for all passes, rather than each pass having its config - is that correct? If so, what does it mean for the configuration to be disabled?

Reply:
This is not entirely correct. This configuration is only concerned with LLVM passes. LLVM passes has two options: Either they are on or they are off. While you could make a separate config class for each LLVM pass, it would always only have one boolean flag hence they are bundled into one configuration.

The static member function `disabled` creates a config where all LLVM passes are disabled and the corresponding member function `isDisabled` checks whether the configuration is an instance of completely disabling all classes.

For our own passes on the other hand, we can pass configurations to them and consequently each of these passes have their own config class and in some cases even subsystems of that pass can have its own config.

Reply:
The disabled property on each pass makes sense, but having it as a property to indicate if all llvm passes are disabled is confusing. I would suggest to drop the isDisabled, and instead of a static member disabled, have a either something like a .WithLlvmPassesDisabled member function, or a constructor that e.g. takes an enum to select predefined configurations via a descriptive name.

Reply:
First of all, notice that `isDisabled` is there because it is used. A disabled configuration for each of the passes does not have any meaning as this already covered by the passes flag by itself. For instance, `alwaysInline()` tells you whether or not the always inline pass will be added or not.

While I think it may make sense to choose a better name for the static function, the proposed change would change the interface when creating a generator from

`Generator(RuleTransformationPassConfiguration::disabled(), LlvmPassesConfiguration::disabled());`

to

`Generator(RuleTransformationPassConfiguration(RuleTransformationPassConfiguration::Config::DisableAll), 
LlvmPassesConfiguration(LlvmPassesConfiguration::Config::DisableAll));`

I am reluctant to make this change as it decreases code readibility substantially and does not provide any additional benefits in terms of readibility or code efficiency. Likewise, checking if a configuration disables a component altogether is currently

```c++
if(config.isDisabled())
{
  
}
```

vs

```c++
if(config.alwaysInline() == false && config.pass2() == false && /* ... */)
{
  
}
```

This will only look worse as more passes are added to the LLVM passes component. Also the readibility of the implementation would be affected by the proposed. If this is something you feel strongly about, lets have a discussion about it.